In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np

In [ ]:
import pickle
pickle_in = open('/content/drive/MyDrive/IIM Nagpur NLP Files/CellPhoneTopicandVotesDF.pickle', 'rb')
df = pickle.load(pickle_in)
df

Topic0  Topic1  Topic2  Topic3  Topic4  Topic5  Topic6  \
Doc0        0.01    0.01    0.01    0.14    0.01    0.01    0.81   
Doc1        0.00    0.07    0.00    0.05    0.12    0.10    0.66   
Doc2        0.04    0.00    0.00    0.36    0.00    0.00    0.58   
Doc3        0.35    0.01    0.13    0.08    0.01    0.16    0.27   
Doc4        0.02    0.09    0.00    0.26    0.14    0.00    0.50   
...          ...     ...     ...     ...     ...     ...     ...   
Doc53625    0.03    0.83    0.03    0.03    0.03    0.03    0.03   
Doc53626    0.07    0.07    0.57    0.07    0.07    0.07    0.07   
Doc53627    0.19    0.69    0.02    0.02    0.02    0.02    0.02   
Doc53628    0.07    0.07    0.07    0.57    0.07    0.07    0.07   
Doc53629    0.10    0.00    0.04    0.00    0.00    0.08    0.77   

          dominant_topic  total_imp_topics custom_topic_name votes  
Doc0                   6                 2               NaN     2  
Doc1                   6                 5               NaN    15  
Doc2                   6                 2               NaN    11  
Doc3                   0                 5          politics     4  
Doc4                   6                 4               NaN    12  
...                  ...               ...               ...   ...  
Doc53625               1                 1           science     0  
Doc53626               2                 7              Math     0  
Doc53627               1                 2           science     0  
Doc53628               3                 7           English     2  
Doc53629               6                 3               NaN     9  

[53630 rows x 11 columns]

In [ ]:
df.loc['Doc43213']

Topic0                                                            0.01
Topic1                                                            0.01
Topic2                                                            0.01
Topic3                                                            0.91
Topic4                                                            0.01
Topic5                                                            0.01
Topic6                                                            0.01
dominant_topic                                                       3
total_imp_topics                                                     1
custom_topic_name                                              English
votes                ['https://images-na.ssl-images-amazon.com/imag...
Name: Doc43213, dtype: object

In [ ]:
df.drop(['Doc43213'], inplace=True)

In [ ]:
df['votes'] = pd.to_numeric(df['votes'])

In [ ]:
df[np.isnan(df['votes']) == True]

Topic0  Topic1  Topic2  Topic3  Topic4  Topic5  Topic6  \
Doc43211    0.06    0.08    0.01    0.11    0.02    0.44    0.28   
Doc43212    0.14    0.14    0.14    0.14    0.14    0.14    0.14   

          dominant_topic  total_imp_topics custom_topic_name  votes  
Doc43211               5                 5               NaN    NaN  
Doc43212               0                 7          politics    NaN

In [ ]:
df.drop(['Doc43211', 'Doc43212'], inplace=True)

In [ ]:
df['votes'].sort_values(ascending=False).head(15)

Doc52646    1737.0
Doc8951     1095.0
Doc8948     1080.0
Doc14710    1074.0
Doc8012      987.0
Doc38348     973.0
Doc52447     946.0
Doc41574     927.0
Doc51330     927.0
Doc42640     813.0
Doc5140      797.0
Doc35580     790.0
Doc1616      718.0
Doc37717     716.0
Doc255       682.0
Name: votes, dtype: float64

In [ ]:
df.drop(['Doc52646'], inplace=True) # removing outlier

In [ ]:
df.reset_index(inplace=True, drop=True)

In [ ]:
df['votes_copy'] = df['votes']
df['votes_copy'].replace(0.0, 1, inplace=True)
df['votelogten'] = np.log10(df['votes_copy'])

In [ ]:
df.head()

Topic0  Topic1  Topic2  Topic3  Topic4  Topic5  Topic6  dominant_topic  \
0    0.01    0.01    0.01    0.14    0.01    0.01    0.81               6   
1    0.00    0.07    0.00    0.05    0.12    0.10    0.66               6   
2    0.04    0.00    0.00    0.36    0.00    0.00    0.58               6   
3    0.35    0.01    0.13    0.08    0.01    0.16    0.27               0   
4    0.02    0.09    0.00    0.26    0.14    0.00    0.50               6   

   total_imp_topics custom_topic_name  votes  votes_copy  votelogten  
0                 2               NaN    2.0         2.0    0.301030  
1                 5               NaN   15.0        15.0    1.176091  
2                 2               NaN   11.0        11.0    1.041393  
3                 5          politics    4.0         4.0    0.602060  
4                 4               NaN   12.0        12.0    1.079181

# Groupby function Inferences

### 1. Total Imp Topics

In [ ]:
g_tot_imp = df.groupby('total_imp_topics')

In [ ]:
# g_tot_imp.get_group(1)

In [ ]:
g_tot_imp.describe()['votes']

count       mean        std  min  25%  50%  75%     max
total_imp_topics                                                           
1                  5326.0   0.507135   4.205367  0.0  0.0  0.0  0.0   159.0
2                 10609.0   1.309171  12.382380  0.0  0.0  0.0  0.0   927.0
3                  8867.0   3.473554  17.637029  0.0  0.0  0.0  2.0   618.0
4                  5957.0   7.479772  36.372746  0.0  0.0  0.0  4.0  1095.0
5                  3140.0  13.235032  57.597245  0.0  0.0  2.0  6.0  1080.0
6                   744.0  19.319892  68.910907  0.0  0.0  3.0  8.0   987.0
7                 18983.0   0.124743   2.265566  0.0  0.0  0.0  0.0   249.0

In [ ]:
for i in range(1,8):
    print(i, '---', g_tot_imp.get_group(i)['votes'].mean())

1 --- 0.5071348103642509
2 --- 1.3091714581958713
3 --- 3.4735536258035413
4 --- 7.4797716971630015
5 --- 13.235031847133758
6 --- 19.31989247311828
7 --- 0.12474319127640521


In [ ]:
df['total_imp_topics'].value_counts()

7    18983
2    10609
3     8867
4     5957
1     5326
5     3140
6      744
Name: total_imp_topics, dtype: int64

**From here, we can see that reviews in which a total of 5 and 6 topics were discussed have the highest mean. But people who discussed 6 topics are very less (744), meanwhile people discussing 5 topics are still enough (3141). <br> Therefore we can conclude that a review which talks about 5 different topics has a higher chance of getting more votes.**

### 2. Dominant Topic


In [ ]:
g_dom = df.groupby('dominant_topic')

In [ ]:
# g_dom.get_group(1)

In [ ]:
g_dom.describe()['votes']

count      mean        std  min  25%  50%  75%     max
dominant_topic                                                          
0                8641.0  0.859970   9.668586  0.0  0.0  0.0  0.0   537.0
1               10486.0  2.067614  16.410887  0.0  0.0  0.0  0.0   797.0
2               13749.0  0.144520   1.031150  0.0  0.0  0.0  0.0    45.0
3                6690.0  3.221973  25.061877  0.0  0.0  0.0  0.0  1080.0
4                 494.0  0.732794   2.557544  0.0  0.0  0.0  0.0    36.0
5                2246.0  6.540071  32.808643  0.0  0.0  0.0  3.0   716.0
6               11320.0  7.291696  38.294358  0.0  0.0  0.0  3.0  1095.0

In [ ]:
for i in range(0,7):
    print(i, '---', g_dom.get_group(i)['votes'].mean())

0 --- 0.8599699108899433
1 --- 2.0676139614724396
2 --- 0.14451960142555823
3 --- 3.2219730941704037
4 --- 0.7327935222672065
5 --- 6.540071237756011
6 --- 7.291696113074205


In [ ]:
df['dominant_topic'].value_counts()

2    13749
6    11320
1    10486
0     8641
3     6690
5     2246
4      494
Name: dominant_topic, dtype: int64

**From here we can conclude Topic 5,6 to be the most important (count of reviews is also decent for 5, 6 is really good). <br>
Overall 6 seems to be the most dominant topic.**

# Linear Regression on Topics and votes (actual values)

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn import linear_model
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
%matplotlib inline

In [ ]:
X = df[df.columns[:7]]
y = df['votes'] #.astype(np.float64)

In [ ]:
X

Topic0  Topic1  Topic2  Topic3  Topic4  Topic5  Topic6
0        0.01    0.01    0.01    0.14    0.01    0.01    0.81
1        0.00    0.07    0.00    0.05    0.12    0.10    0.66
2        0.04    0.00    0.00    0.36    0.00    0.00    0.58
3        0.35    0.01    0.13    0.08    0.01    0.16    0.27
4        0.02    0.09    0.00    0.26    0.14    0.00    0.50
...       ...     ...     ...     ...     ...     ...     ...
53621    0.03    0.83    0.03    0.03    0.03    0.03    0.03
53622    0.07    0.07    0.57    0.07    0.07    0.07    0.07
53623    0.19    0.69    0.02    0.02    0.02    0.02    0.02
53624    0.07    0.07    0.07    0.57    0.07    0.07    0.07
53625    0.10    0.00    0.04    0.00    0.00    0.08    0.77

[53626 rows x 7 columns]

In [ ]:
y

0         2.0
1        15.0
2        11.0
3         4.0
4        12.0
         ... 
53621     0.0
53622     0.0
53623     0.0
53624     2.0
53625     9.0
Name: votes, Length: 53626, dtype: float64

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.3, random_state = 0)

In [ ]:
regressor = LinearRegression()

In [ ]:
regressor.fit(X_train, y_train)

LinearRegression()

In [ ]:
regressor.coef_

array([24.23585382, 25.02916015, 21.19266673, 27.58935605, 23.37780799,
       38.52771195, 33.04275102])

In [ ]:
regressor.intercept_

-24.228666244968956

In [ ]:
y_pred = regressor.predict(X_test)      # testing our model using X test data
y_pred

array([0.41251092, 7.49972928, 1.41676266, ..., 6.66183598, 2.79067683,
       5.8023808 ])

In [ ]:
y_test

36521    0.0
51209    3.0
38985    0.0
45853    4.0
20010    0.0
        ... 
19914    0.0
45873    3.0
39711    5.0
21613    0.0
18955    0.0
Name: votes, Length: 16088, dtype: float64

In [ ]:
regressor.score(X_test, y_test)

0.01508938846865171

# Linear Regression on Topics and votelogten

In [ ]:
X = df[df.columns[:7]]
y = df['votelogten'] #.astype(np.float64)

In [ ]:
y

0        0.301030
1        1.176091
2        1.041393
3        0.602060
4        1.079181
           ...   
53621    0.000000
53622    0.000000
53623    0.000000
53624    0.301030
53625    0.954243
Name: votelogten, Length: 53626, dtype: float64

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.3, random_state = 0)

In [ ]:
regressor = LinearRegression()

In [ ]:
regressor.fit(X_train, y_train)

LinearRegression()

In [ ]:
regressor.coef_

array([0.79923348, 0.86652771, 0.63466619, 0.91512446, 0.69855422,
       1.27840412, 1.15011126])

In [ ]:
regressor.intercept_

-0.7484161918586559

In [ ]:
y_pred = regressor.predict(X_test)      # testing our model using X test data
y_pred

array([0.04190177, 0.31924863, 0.09620898, ..., 0.2904948 , 0.13955081,
       0.27062294])

In [ ]:
y_test

36521    0.000000
51209    0.477121
38985    0.000000
45853    0.602060
20010    0.000000
           ...   
19914    0.000000
45873    0.477121
39711    0.698970
21613    0.000000
18955    0.000000
Name: votelogten, Length: 16088, dtype: float64

In [ ]:
regressor.score(X_test, y_test)

0.10417717620690203

In [ ]:
# pickle_out = open("CellPhoneTopicandVotesDF(votelogten_and_outliers_removed).pickle", 'wb')
# pickle.dump(df, pickle_out)
# pickle_out.close()